### ISSUE TITLE
Cannot run Invoke-D365SDPInstall twice in a row: "Stopping because some assembly (DLL) files seems to be loaded ..."

### ISSUE THREAD

**Issue by devax** [d365collaborative/d365fo.tools#241]
Hi guys,

I might be missing something here, but since version 0.5.11 my deployment scripts are broken. If I use version 0.5.11 and above, first invocation of `Invoke-D365SDPInstall` runs fine, but subsequent invocations will output a warning but the command will not execute.

So here is an example together with the output:
```
PS C:\windows\system32> # First invocation
Invoke-D365SDPInstall -Path "C:\agent\_work\r1\a\_unpacked" -Command SetTopology

PS C:\windows\system32>
PS C:\windows\system32> # Subsequent invocations
Invoke-D365SDPInstall -Path "C:\agent\_work\r1\a\_unpacked" -Command SetTopology
WARNING: [12:36:48][Test-AssembliesLoaded] Stopping because some assembly (DLL) files seems to be loaded into memory.
[12:36:48][Invoke-D365SDPInstall] It seems that you have executed some cmdlets that required to load some Dynamics 356 Finance & Operations assemblies into memory. Please close and restart you PowerShell session / console, and start a fresh. Please note that you should execute the failed command immediately after importing the module.

PS C:\windows\system32>
```

The above example obviously does not make a lot of sense. My real life script does more look like this:
```
Invoke-D365SDPInstall -Path $myPath -Command SetTopology
Invoke-D365SDPInstall -Path $myPath -Command Generate -RunbookId $runbookId
Invoke-D365SDPInstall -Path $myPath -Command Import -RunbookId $runbookId
Invoke-D365SDPInstall -Path $myPath -Command Execute -RunbookId $runbookId
```
Same behavior here: Only first invocation of the command will work correctly, subsequent invocations will only output the warning but not do anything.

What makes things even worse is that the warning does not stop the script from executing. So the script will continue to run and - on first sight - the deployment will look to be successful, when it actually was not.

For now I just stick with version 0.5.9 that works with my scripts.

Am I the only one experiencing this kind of problems? Is this new behavior design intended? 
Sorry if I'm missing something elementary here. Thanks.

**Comment by Splaxi** [comment]
First of - thanks for posting the issue.

It is actually on purpose we implemented this new check. We had other issues where the module had loaded some dll files, that the update we tried to apply, had to update/replace. So we ended up with not being able to run the update, before exiting the session and then re-running the update. That made us try and implement this check, to protect others from the same error and mistake.

May I ask why you simply don't run the `-Command RunAll` to install the desired update? This will do all 4 steps for you, without you having to worry.

We use the `-Command RunAll` and that is the reason that we never found this error.

With that said, I need to dig into things a bit more and see if I can way to support both scenarios. 

Can you share some details about the package / update you are trying to apply? Is it a package from a build or a MS package?

**Comment by devax** [comment]
Thank you for your answer -- much appreciated!
I'm installing custom build packages.
I was not aware that `-Command RunAll` is what people should use. I am relatively new to deployment packages management and I'm under the impression that the generally available documentation is not very detailed. Specifically, I was under the impression that I had to create a new runbook ID for every deployment. I'm not sure what will happen when a deployment does not suceed (e.g. synchronization errors) and in that state a new deployment will be started using the same runbook ID. But to be honest, I'm also uncertain what will happen in the same scenario when different runbook IDs are used.
It just felt being a bit more in control when creating new runbook IDs with each deployment.

So you guys have no issues re-using the same runbook ID? If not, I can also try to switch over to use "RunAll".

**Comment by Splaxi** [comment]
Are you talking about the documentation provided by Microsoft for their tools?

Or our tools? 😀 

We expect some kind of knowledge, simply because we build on top of the tools that are provided by Microsoft. But we also try to remove some of the boring details, with all the default values for each function / cmdlet. 

Let us take things from the top. 

You have 3 different options that simply just installs the package for you and where you don't have to do anything. 

`-QuickInstallAll`
This just runs the installation and *doesn't* care about the topology file. It should install all components. Requires administrative privileges. 

http://yetanotherdynamicsaxblog.blogspot.com/2018/02/installing-software-deployable-package.html?m=1

`-DevInstall`
The name gives it away, it is meant for installing the package on a MS hosted, non-admin box. This should also just work, but I never fully understood if it will skip the topology file or you need to update the topology file first. It doesn't run dbsync, you need to do that yourself. 

https://docs.microsoft.com/en-us/dynamics365/unified-operations/dev-itpro/deployment/install-deployable-package

`-Command RunAll`
This is the magic bullet! It simply runs all the steps that you are executing by hand, so it takes all 4 executions and makes it one. 

For the `-QuickInstallAll` and `-DevInstall` you can't give a runbookid, the Axupdateinstaller.exe will simply do that for you. 

For the `-Command RunAll` you can specify a runbookid. 

The runbookid has 2 purposes, one is to separate different installations from each other and the last is to make sure that you can re-run or skip steps of your installation after it has failed and you can restart the installation. 

At the moment we, apparently, don't support re-run, without having to restart the powershell session. As I said, I will be looking into that and see if I can find a solution that works for all scenarios. 

Let me know if you need anything else 😀

**Comment by Splaxi** [comment]
And I think I found the bug!

Please download the 0.5.21 release in like 10-15 minutes and test that things work like you expect.

**Comment by devax** [comment]
I had Microsoft's documentation in mind when I was slightly complaining about what is available. I totally understand that it is out of scope for this project to fill this gap (if there is any; it might be my personal inability to find the right places to read things up). Nevertheless, thank you very much that you have laid out the available options to me. This totally cleared things up a bit for me.

I have tested your fix with the currently newest version 0.5.22 and I can confirm, it works well with my existing scripts! Thank you very much for addressing this issue.

For now, I will keep my existing scripts in place, but I will experiment with the `-Command RunAll` option when I will be working on the deployment scripts next time.

Please keep up the excellent work!

**Comment by Splaxi** [comment]
I'm glad that we were able to fix the bug!

Do you have Twitter? Please ping me on Twitter, and I'm be happy help with any issues you might face, even things outside what we support with the tools.

I hope the tools make your daily life easier, and I hope you will provide feedback on things we should focus on for the road map, so we can make things even better.
